### PR TITLE
fix(models): Handle empty message in LiteLLM response (fixes #3618)

### LINKED ISSUE #3618: ValueError("No message in response") when turn ends on tool call
**Describe the bug**
When using `google-adk` with LiteLLM and tools, a run can end with a tool call/ tool response and **no final agent message**. In these cases, the LiteLLM wrapper raises `ValueError: No message in response` from `_model_response_to_generate_content_response`. 

Technically, I can catch this error in my own code, but I'd like to understand if this behaviour is **intended**. Shouldn't a turn that ends only with a tool call/tool response be considered a valid outcome for ADK? If yes, then raising an error in this case feels wrong.

This error is not easy to prevent, since the LLM model can act like that sometimes, by writing the "final" message as an intermediate response and finalizing the turn by calling a tool, even if instructed not to. It is also annoying to integrate with other frameworks like **Phoenix Otel**, since this span comes as an error, even if the system is supposed to work like that.

**To Reproduce**
1. Create an agent that uses a LiteLLM-backed model with tools enabled.
2. Run the agent via `Runner.run_async(...)` on a query that triggers a tool call and ends the turn without a **final** agent message. Note that in this case we could have intermediate responses. What matters is that the last events should be a tool call and tool response.
3. Observe that `_model_response_to_generate_content_response` inspects the LiteLLM response, and then raises `ValueError: No message in response`

**Expected behavior**
I'd expect a turn that ends with a tool response (with no final agent message)  to be treated as a **valid outcome**, not as a fatal error.

If the response has no message, what ADK could do is to construct an `LlmResponse` with an empty text but valid tool/actions and metadata, and avoid raising `ValueError` in this case.

**Desktop (please complete the following information):**
 - OS: Linux (WSL2)
 - Python version: 3.10
 - ADK version: 1.18.0
 - openinference-instrumentation-google-adk: 0.1.6
 - arize-phoenix-otel: 0.13.1

 **Model Information:**
 - Are you using LiteLLM: Yes
 - Which model is being used: gemini-2.5-flash

**Additional context**
Usage pattern:  multi-agent Runner.run_async(...) with tools and Otel tracing enabled.

### PR DESCRIPTION
**Pull request by sarojrout**
When a turn ends with only tool calls and no final agent message, _model_response_to_generate_content_response was raising ValueError. This fix returns an empty LlmResponse instead, allowing workflows to continue gracefully.

Changes:
- Return empty LlmResponse when message is None or empty
- Preserve finish_reason and usage_metadata if available
- Added test cases for the edge cases

**Please ensure you have read the [contribution guide](https://github.com/google/adk-python/blob/main/CONTRIBUTING.md) before creating a pull request.**

### Link to Issue or Description of Change

**1. Link to an existing issue (if applicable):**

- Closes: #3618 
- Related: #3618

**2. Or, if no issue exists, describe the change:**

_If applicable, please follow the issue templates to provide as much detail as
possible._

**Problem:**
When using `google-adk` with LiteLLM and tools, a run can end with a tool call/tool response and **no final agent message**. In these cases, the LiteLLM wrapper was raising `ValueError: No message in response` from `_model_response_to_generate_content_response`, causing workflows to crash even though this is a valid scenario.

This is particularly problematic in multi-agent workflows where:
- A turn can legitimately end with only tool calls (no text message)
- The LLM model may write intermediate responses and finalize the turn by calling a tool
- Integration with observability frameworks marks these as errors even though the system is working correctly

**Solution:**
Instead of raising `ValueError` when `message` is `None` or empty, the code now:
1. Creates an empty `LlmResponse` with empty content parts
2. Preserves `finish_reason` if available (properly mapped to the appropriate enum)
3. Preserves `usage_metadata` if available
4. Returns the response gracefully, allowing workflows to continue

This solution treats a turn ending with only tool calls as a **valid outcome**, which aligns with the expected behavior described in the issue. The fix is backward-compatible - existing code that works will continue to work, and the edge case is now handled gracefully.

### Testing Plan

_Please describe the tests that you ran to verify your changes. This is required
for all PRs that are not small documentation or typo fixes._

**Unit Tests:**

- [x] I have added or updated unit tests for my change.
- [x] All unit tests pass locally.

Added three new test cases in `tests/unittests/models/test_litellm.py`:

1. `test_model_response_to_generate_content_response_no_message_with_finish_reason`
   - Tests response with no message but has `finish_reason="tool_calls"`
   - Verifies empty LlmResponse is returned with correct finish_reason and usage_metadata

2. `test_model_response_to_generate_content_response_no_message_no_finish_reason`
   - Tests response with no message and no finish_reason
   - Verifies empty LlmResponse is returned without errors

3. `test_model_response_to_generate_content_response_empty_message_dict`
   - Tests response with empty message dict (`message = {}`)
   - Verifies empty LlmResponse is returned correctly

**Test Results:**
```

$ pytest tests/unittests/models/test_litellm.py -k "test_model_response_to_generate_content_response_no_message" -v

tests/unittests/models/test_litellm.py::test_model_response_to_generate_content_response_no_message_with_finish_reason PASSED [ 50%]
tests/unittests/models/test_litellm.py::test_model_response_to_generate_content_response_no_message_no_finish_reason PASSED [100%]
================================ 2 passed, 102 deselected in 5.69s ================================
tests/unittests/models/test_litellm.py::test_model_response_to_generate_content_response_empty_message_dict PASSED [100%]

================================ 1 passed, 103 deselected in 5.71s ==========================
```

**All existing tests pass:**
```bash
$ pytest tests/unittests/models/test_litellm.py -v
=============================104 passed, 4 warnings in 6.40s =================================```
```
**Manual End-to-End (E2E) Tests:**

I have tested with the below utility code which you can try to test out

```
from litellm import ChatCompletionAssistantMessage
from litellm.types.utils import ModelResponse
from google.adk.models.lite_llm import _model_response_to_generate_content_response

print("=" * 60)

# Test 1: Response WITHOUT message (the bug scenario)
print("\n Test 1: Response WITHOUT message (bug scenario)")
print("-" * 60)
response_no_message = ModelResponse(
    model="test",
    choices=[{
        "finish_reason": "tool_calls",
        # message is missing - this used to raise ValueError
    }],
    usage={
        "prompt_tokens": 10,
        "completion_tokens": 5,
        "total_tokens": 15,
    },
)

try:
    result1 = _model_response_to_generate_content_response(response_no_message)
    print("SUCCESS: No ValueError raised!")
    print(f"   finish_reason: {result1.finish_reason}")
    print(f"   content parts: {len(result1.content.parts)}")
    print(f"   usage tokens: {result1.usage_metadata.total_token_count}")
    print("  Empty LlmResponse returned correctly")
except ValueError as e:
    if "No message in response" in str(e):
        print(f"FAILED: Still raises ValueError: {e}")
        print("   The fix is NOT working!")
    else:
        raise

# Test 2: Response WITH message (normal case - should still work)
print("\n Test 2: Response WITH message (normal case)")
print("-" * 60)
response_with_message = ModelResponse(
    model="test",
    choices=[{
        "message": ChatCompletionAssistantMessage(
            role="assistant",
            content="This is a normal response with text",
        ),
        "finish_reason": "stop",
    }],
    usage={
        "prompt_tokens": 10,
        "completion_tokens": 8,
        "total_tokens": 18,
    },
)

try:
    result2 = _model_response_to_generate_content_response(response_with_message)
    print("SUCCESS: Normal case still works!")
    print(f"   finish_reason: {result2.finish_reason}")
    print(f"   content parts: {len(result2.content.parts)}")
    if result2.content.parts:
        print(f"   text content: {result2.content.parts[0].text}")
    print(f"   usage tokens: {result2.usage_metadata.total_token_count}")
    print(" Normal message processing works correctly")
except Exception as e:
    print(f"FAILED: Error in normal case: {e}")
    raise

# Test 3: Response WITH message AND tool calls
print("\n Test 3: Response WITH message AND tool calls")
print("-" * 60)
from litellm import ChatCompletionMessageToolCall, Function

response_with_tool_calls = ModelResponse(
    model="test",
    choices=[{
        "message": ChatCompletionAssistantMessage(
            role="assistant",
            content="I'll call a tool for you",
            tool_calls=[
                ChatCompletionMessageToolCall(
                    type="function",
                    id="call_123",
                    function=Function(
                        name="test_function",
                        arguments='{"arg": "value"}',
                    ),
                )
            ],
        ),
        "finish_reason": "tool_calls",
    }],
    usage={
        "prompt_tokens": 10,
        "completion_tokens": 12,
        "total_tokens": 22,
    },
)

try:
    result3 = _model_response_to_generate_content_response(response_with_tool_calls)
    print("SUCCESS: Tool calls case works!")
    print(f"   finish_reason: {result3.finish_reason}")
    print(f"   content parts: {len(result3.content.parts)}")
    for i, part in enumerate(result3.content.parts):
        if part.text:
            print(f"   part {i}: text = '{part.text}'")
        if part.function_call:
            print(f"   part {i}: function_call = {part.function_call.name}")
    print(f"   usage tokens: {result3.usage_metadata.total_token_count}")
    print("Tool calls processing works correctly")
except Exception as e:
    print(f"FAILED: Error in tool calls case: {e}")
    raise

print("\n" + "=" * 60)
```

### Checklist

- [x] I have read the [CONTRIBUTING.md](https://github.com/google/adk-python/blob/main/CONTRIBUTING.md) document.
- [x] I have performed a self-review of my own code.
- [x] I have commented my code, particularly in hard-to-understand areas.
      - Added docstring explaining the behavior when message is None
      - Added inline comments explaining the fix logic
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] New and existing unit tests pass locally with my changes.
        - All 104 existing tests pass
        - All 3 new tests pass
- [x] I have manually tested my changes end-to-end.
- [x] Any dependent changes have been merged and published in downstream modules.
    - No dependent changes required

### Additional context

**Code Changes:**
- Modified `src/google/adk/models/lite_llm.py::_model_response_to_generate_content_response()`
- Changed from raising `ValueError` to returning empty `LlmResponse` when message is None/empty
- Preserves all available metadata (finish_reason, usage_metadata) in the empty response

### CODE REVIEW

File: src/google/adk/models/lite_llm.py [hunk 1]
```diff
@@ -837,8 +841,35 @@ def _model_response_to_generate_content_response(
...
+        )
+    if response.get("usage", None):
+      llm_response.usage_metadata = types.GenerateContentResponseUsageMetadata(
+          prompt_token_count=response["usage"].get("prompt_tokens", 0),
+          candidates_token_count=response["usage"].get("completion_tokens", 0),
+          total_token_count=response["usage"].get("total_tokens", 0),
+          cached_content_token_count=_extract_cached_prompt_tokens(
+              response["usage"]
+          ),
+      )
```
[the remark this answers is not in the document]
Comment (sarojrout) [reply src/google/adk/models/lite_llm.py]: pushed by removing the duplicate code and replaced with _convert_reasoning_value_to_parts() can you re-verify? @copilot

**Comment by rohityan** [comment]
Hi @sarojrout ,Thank you for your work on this pull request. We appreciate the effort you've invested.
Before we can proceed with the review can you please fix the lint errors. You can use autoformat.sh.

**Comment by sarojrout** [comment]
> Hi @sarojrout ,Thank you for your work on this pull request. We appreciate the effort you've invested. Before we can proceed with the review can you please fix the lint errors. You can use autoformat.sh.

sure will do it in sometime. thanks for reviewing @ryanaiagent

**Comment by sarojrout** [comment]
> > Hi @sarojrout ,Thank you for your work on this pull request. We appreciate the effort you've invested. Before we can proceed with the review can you please fix the lint errors. You can use autoformat.sh.
> 
> sure will do it in sometime. thanks for reviewing @ryanaiagent

pls check @ryanaiagent whenever you get time. thanks!

**Comment by rohityan** [comment]
Hi @sarojrout, Your PR has been received by the team and is currently under review. We will provide feedback as soon as we have an update to share.

**Comment by rohityan** [comment]
Hi @ankursharmas , can you please review this.

**Comment by FreshYop** [comment]
Any news on this issue ? I got the same errors :)

**Comment by sarojrout** [comment]
@ryanaiagent , any updates on this PR?

**Comment by rohityan** [comment]
Hi @sarojrout , We're still looking into this, so we appreciate your patience and support.

**Comment by pattabhia** [comment]
when is this fix planned for release?
